pkg/scalers/kafka/kafka_scaler_oauth_token_provider.go

```diff
@@ -18,36 +18,94 @@ package kafka
 
 import (
 	"context"
+	"sync"
+	"time"
 
 	"github.com/IBM/sarama"
+	"github.com/aws/aws-msk-iam-sasl-signer-go/signer"
+	"github.com/aws/aws-sdk-go-v2/aws"
 	"golang.org/x/oauth2"
 	"golang.org/x/oauth2/clientcredentials"
 )
 
-type TokenProvider struct {
+type TokenProvider interface {
+	sarama.AccessTokenProvider
+	String() string
+}
+
+type oauthBearerTokenProvider struct {
 	tokenSource oauth2.TokenSource
 	extensions  map[string]string
 }
 
-func OAuthBearerTokenProvider(clientID, clientSecret, tokenURL string, scopes []string, extensions map[string]string) sarama.AccessTokenProvider {
+func OAuthBearerTokenProvider(clientID, clientSecret, tokenURL string, scopes []string, extensions map[string]string) TokenProvider {
 	cfg := clientcredentials.Config{
 		ClientID:     clientID,
 		ClientSecret: clientSecret,
 		TokenURL:     tokenURL,
 		Scopes:       scopes,
 	}
 
-	return &TokenProvider{
+	return &oauthBearerTokenProvider{
 		tokenSource: cfg.TokenSource(context.Background()),
 		extensions:  extensions,
 	}
 }
 
-func (t *TokenProvider) Token() (*sarama.AccessToken, error) {
-	token, err := t.tokenSource.Token()
+func (o *oauthBearerTokenProvider) Token() (*sarama.AccessToken, error) {
+	token, err := o.tokenSource.Token()
 	if err != nil {
 		return nil, err
 	}
 
-	return &sarama.AccessToken{Token: token.AccessToken, Extensions: t.extensions}, nil
+	return &sarama.AccessToken{Token: token.AccessToken, Extensions: o.extensions}, nil
+}
+
+func (o *oauthBearerTokenProvider) String() string {
+	return "OAuthBearer"
+}
+
+// mskTokenJitter is the amount of time to subtract from the token expiration time to ensure the token is refreshed before it expires.
+const mskTokenJitter = 30 * time.Second
```

Comment:
Thanks for pointing it too, after digging, I found the following:

* There's this from the Java client Kafka implementation:

https://github.com/apache/kafka/blob/a1ca788c993e3b217a0904c3f4768f0682dc840b/clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java#L683-L700

```java
 // pick a random percentage between 85% and 95% for session re-authentication
double pctWindowFactorToTakeNetworkLatencyAndClockDriftIntoAccount = 0.85;
double pctWindowJitterToAvoidReauthenticationStormAcrossManyChannelsSimultaneously = 0.10;
double pctToUse = pctWindowFactorToTakeNetworkLatencyAndClockDriftIntoAccount + RNG.nextDouble()
                        * pctWindowJitterToAvoidReauthenticationStormAcrossManyChannelsSimultaneously;
```

And the [same code appears in Sarama](https://github.com/IBM/sarama/blob/4ad35041300e1c15ba58b630745ac8eb05f30c10/broker.go#L1590-L1605) but it actually requires the response to contain the expiration, and this is only when using a SaslAuthenticateRequest > v0 which is why we need to override the Kafka version used to at least [v2.2.0](https://cwiki.apache.org/confluence/display/KAFKA/KIP-368%3A+Allow+SASL+Connections+to+Periodically+Re-Authenticate). With this override, the SaslAuthenticateResponse contains the SessionLifetimeMs which is used later and all is good.

I think this is something that will definitely need to be mentioned in the documentation. Otherwise, the token will not be refreshed until the broker closes the connection.

Reply:
I am still deep diving on this mechanism.

For example, the `res.SessionLifetimeMs` from my MSK cluster is ~43200000ms or 12 hours. Which doesn't seem to match the expiry of the token which is 5 minutes. I am reading https://cwiki.apache.org/confluence/display/KAFKA/KIP-368%3A+Allow+SASL+Connections+to+Periodically+Re-Authenticate

Reply:
Since this is a broker/client mechanism unrelated to Keda, I've removed this jitter parameter 👍